src/System.Text.Json/System/Text/Json/JsonReader.cs

```diff
@@ -1,383 +1,836 @@
+        /// <see cref="JsonTokenType.PropertyName" />. Otherwise, this value should be set to
+        /// <see cref="ReadOnlySpan{T}.Empty"/>.
+        /// </summary>
+        public ReadOnlySpan<byte> Value { get; private set; }
+
+        /// <summary>
+        /// Gets the JSON value type of the last processed token. The contents of this
+        /// is only relevant when <see cref="TokenType" /> is <see cref="JsonTokenType.Value" /> or
+        /// <see cref="JsonTokenType.PropertyName" />.
+        /// </summary>
+        public JsonValueType ValueType { get; private set; }
+
+        /// <summary>
+        /// Gets the encoder instance used when the reader was constructed.
+        /// </summary>
+        public TextEncoder Encoder => _encoder;
+
+        /// <summary>
+        /// Constructs a new JsonReader instance. This is a stack-only type.
+        /// </summary>
+        /// <param name="data">The <see cref="Span{byte}"/> value to consume. </param>
+        /// <param name="encoder">An encoder used for decoding bytes from <paramref name="data"/> into characters.</param>
+        public JsonReader(ReadOnlySpan<byte> data, TextEncoder encoder)
         {
-            // Start = 0 state reserved for internal use
-            ObjectStart = 1,
-            ObjectEnd = 2,
-            ArrayStart = 3,
-            ArrayEnd = 4,
-            Property = 5,
-            Value = 6
-        };
-
-        public enum JsonValueType
-        {
-            String,
-            Number,
-            Object,
-            Array,
-            True,
-            False,
-            Null
+            _buffer = data;
+            _encoder = encoder;
+            _depth = 0;
+            _containerMask = 0;
+
+            if (encoder.IsInvariantUtf8)
+                _encoderState = JsonEncoderState.UseFastUtf8;
+            else if (encoder.IsInvariantUtf16)
+                _encoderState = JsonEncoderState.UseFastUtf16;
+            else
+                _encoderState = JsonEncoderState.UseFullEncoder;
+
+            TokenType = JsonTokenType.None;
+            Value = ReadOnlySpan<byte>.Empty;
+            ValueType = JsonValueType.Unknown;
         }
 
-        public JsonReader(Utf8String str)
+        /// <summary>
+        /// Read the next token from the data buffer.
+        /// </summary>
+        /// <returns>True if the token was read successfully, else false.</returns>
+        public bool Read()
         {
-            _str = str.Trim();
-            _index = 0;
-            _insideObject = 0;
-            _insideArray = 0;
-            TokenType = 0;
-            _jsonStartIsObject = (byte)_str[0] == '{';
+            ref byte bytes = ref _buffer.DangerousGetPinnableReference();
+            int count = _buffer.Length;
+            int skip = SkipWhiteSpace(ref bytes, count);
+
+            ref byte next = ref Unsafe.Add(ref bytes, skip);
+            count -= skip;
+
+            int step = GetNextCharAscii(ref next, count, out char ch);
```

Comment:
This function is only used for reading characters that result in state changes (same as Control characters in the writer). I could probably rename this or add a comment to reflect this.